File: app/src/main/java/com/dumuzeyn/mp3player/FullPlayerPagerController.java
package com.dumuzeyn.mp3player;

import android.graphics.drawable.GradientDrawable;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager2.widget.ViewPager2;

/** Caches three pages and maps the requested inverse swipe order. */
final class FullPlayerPagerController implements AutoCloseable {
    private final MainActivityCore host;
    private final FullPlayerPlaybackPage playerPage;
    private final LyricsPageController lyricsPage;
    private final QueuePageController queuePage;
    private final View[] segments = new View[3];
    private ViewPager2 pager;
    private int selected = FullPlayerPageOrder.PLAYER;
    private boolean hostVisible = true;

    FullPlayerPagerController(MainActivityCore host, PlaybackActions actions,
            PlaybackStateProvider playbackState) {
        this.host = host;
        playerPage = new FullPlayerPlaybackPage(host, actions, playbackState);
        lyricsPage = new LyricsPageController(host, playbackState);
        queuePage = new QueuePageController(host, playbackState);
    }

    View createPager() {
        pager = new ViewPager2(host);
        pager.setOrientation(ViewPager2.ORIENTATION_HORIZONTAL);
        pager.setOffscreenPageLimit(1);
        pager.setAdapter(new PagesAdapter());
        pager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override public void onPageSelected(int position) {
                selected = position;
                updateActivePage();
                updateIndicator();
            }
        });
        pager.setCurrentItem(FullPlayerPageOrder.PLAYER, false);
        return pager;
    }

    View createIndicator() {
        LinearLayout row = new LinearLayout(host);
        row.setGravity(Gravity.CENTER);
        for (int logical = 0; logical < 3; logical++) {
            int position = 2 - logical;
            View segment = new View(host);
            segments[position] = segment;
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                    host.dp(22), host.dp(4));
            params.setMargins(host.dp(3), 0, host.dp(3), 0);
            row.addView(segment, params);
        }
        updateIndicator();
        return row;
    }

    void refresh() {
        playerPage.refresh(true);
        if (selected == FullPlayerPageOrder.LYRICS) {
            lyricsPage.refresh();
        } else if (selected == FullPlayerPageOrder.QUEUE) {
            queuePage.refresh();
        }
    }

    void setHostVisible(boolean visible) {
        hostVisible = visible;
        updateActivePage();
    }

    int selectedPage() {
        return selected;
    }

    private void updateActivePage() {
        playerPage.setActive(hostVisible && selected == FullPlayerPageOrder.PLAYER);
        lyricsPage.setActive(hostVisible && selected == FullPlayerPageOrder.LYRICS);
        queuePage.setActive(hostVisible && selected == FullPlayerPageOrder.QUEUE);
    }

    private void updateIndicator() {
        for (int position = 0; position < segments.length; position++) {
            View segment = segments[position];
            if (segment == null) {
                continue;
            }
            GradientDrawable background = new GradientDrawable();
            background.setCornerRadius(host.dp(3));
            background.setColor(position == selected ? host.purple : host.cardStroke);
            segment.setBackground(background);
        }
    }

    @Override public void close() {
        playerPage.close();
        lyricsPage.close();
        queuePage.close();
        if (pager != null) {
            pager.setAdapter(null);
            pager = null;
        }
    }

    private final class PagesAdapter extends RecyclerView.Adapter<PageHolder> {
        @NonNull @Override public PageHolder onCreateViewHolder(
                @NonNull ViewGroup parent, int viewType) {
            FrameLayout container = new FrameLayout(host);
            container.setLayoutParams(new RecyclerView.LayoutParams(-1, -1));
            return new PageHolder(container);
        }

        @Override public void onBindViewHolder(@NonNull PageHolder holder, int position) {
            holder.container.removeAllViews();
            View page = position == FullPlayerPageOrder.PLAYER ? playerPage.createView()
                    : position == FullPlayerPageOrder.LYRICS ? lyricsPage.createView()
                    : queuePage.createView();
            if (page.getParent() instanceof ViewGroup) {
                ((ViewGroup) page.getParent()).removeView(page);
            }
            holder.container.addView(page, new FrameLayout.LayoutParams(-1, -1));
        }

        @Override public int getItemCount() {
            return 3;
        }
    }

    static final class PageHolder extends RecyclerView.ViewHolder {
        final FrameLayout container;
        PageHolder(FrameLayout container) {
            super(container);
            this.container = container;
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/FullPlayerPagerController.kt
package com.dumuzeyn.mp3player

import android.graphics.drawable.GradientDrawable
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.widget.ViewPager2

/** Caches the player, lyrics, and queue pages in their visual swipe order. */
internal class FullPlayerPagerController(
    private val host: MainActivityCore,
    actions: PlaybackActions,
    playbackState: PlaybackStateProvider,
) : AutoCloseable {
    private val playerPage = FullPlayerPlaybackPage(host, actions, playbackState)
    private val lyricsPage = LyricsPageController(host, playbackState)
    private val queuePage = QueuePageController(host, playbackState) {
        showPage(FullPlayerPageOrder.LYRICS)
    }
    private val segments = arrayOfNulls<View>(3)
    private var pager: ViewPager2? = null
    private var selected = FullPlayerPageOrder.PLAYER
    private var hostVisible = true

    fun createPager(): View {
        val created = ViewPager2(host).apply {
            orientation = ViewPager2.ORIENTATION_HORIZONTAL
            offscreenPageLimit = 1
            adapter = PagesAdapter()
            registerOnPageChangeCallback(
                object : ViewPager2.OnPageChangeCallback() {
                    override fun onPageSelected(position: Int) {
                        selected = position
                        updateActivePage()
                        updateIndicator()
                    }
                },
            )
            setCurrentItem(FullPlayerPageOrder.PLAYER, false)
        }
        pager = created
        return created
    }

    fun createIndicator(): View {
        val row = LinearLayout(host).apply { gravity = Gravity.CENTER }
        for (position in segments.indices) {
            val segment = View(host)
            segments[position] = segment
            val target = FrameLayout(host).apply {
                isClickable = true
                isFocusable = true
                contentDescription = pageDescription(position)
                setOnClickListener { showPage(position) }
                addView(
                    segment,
                    FrameLayout.LayoutParams(host.dp(22), host.dp(4), Gravity.CENTER),
                )
            }
            row.addView(target, LinearLayout.LayoutParams(host.dp(48), host.dp(18)))
        }
        updateIndicator()
        return row
    }

    fun refresh() {
        playerPage.refresh(true)
        when (selected) {
            FullPlayerPageOrder.LYRICS -> lyricsPage.refresh()
            FullPlayerPageOrder.QUEUE -> queuePage.refresh()
        }
    }

    fun setHostVisible(visible: Boolean) {
        hostVisible = visible
        updateActivePage()
    }

    fun selectedPage(): Int = selected

    private fun updateActivePage() {
        playerPage.setActive(hostVisible && selected == FullPlayerPageOrder.PLAYER)
        lyricsPage.setActive(hostVisible && selected == FullPlayerPageOrder.LYRICS)
        queuePage.setActive(hostVisible && selected == FullPlayerPageOrder.QUEUE)
    }

    private fun updateIndicator() {
        segments.forEachIndexed { position, segment ->
            if (segment == null) return@forEachIndexed
            segment.background = GradientDrawable().apply {
                cornerRadius = host.dp(3).toFloat()
                setColor(if (position == selected) host.purple else host.cardStroke)
            }
        }
    }

    private fun pageDescription(position: Int): String = when (position) {
        FullPlayerPageOrder.PLAYER -> host.tr("Open player", "Открыть плеер")
        FullPlayerPageOrder.LYRICS -> host.tr("Open lyrics", "Открыть текст")
        else -> host.tr("Open queue", "Открыть очередь")
    }

    private fun showPage(position: Int) {
        pager?.setCurrentItem(position, host.appearanceState.animations)
    }

    override fun close() {
        playerPage.close()
        lyricsPage.close()
        queuePage.close()
        pager?.adapter = null
        pager = null
    }

    private inner class PagesAdapter : RecyclerView.Adapter<PageHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PageHolder {
            val container = FrameLayout(host).apply {
                layoutParams = RecyclerView.LayoutParams(-1, -1)
            }
            return PageHolder(container)
        }

        override fun onBindViewHolder(holder: PageHolder, position: Int) {
            holder.container.removeAllViews()
            val page = when (position) {
                FullPlayerPageOrder.PLAYER -> playerPage.createView()
                FullPlayerPageOrder.LYRICS -> lyricsPage.createView()
                else -> queuePage.createView()
            }
            (page.parent as? ViewGroup)?.removeView(page)
            holder.container.addView(page, FrameLayout.LayoutParams(-1, -1))
        }

        override fun getItemCount(): Int = 3
    }

    class PageHolder(val container: FrameLayout) : RecyclerView.ViewHolder(container)
}
